### ISSUE TITLE
kopf run does not work with click 8.x

### ISSUE THREAD

**Issue by pshchelo** [nolar/kopf#767]
## New click breaks kopf

New `click` 8.0 has been released on May 11, 2021
With it, the latest kopf release fails to start the `kopf run`, seems like something changed in the handling of defaults for logging configuration

## Description

<details><summary>The exact command to reproduce the issue</summary>

```bash
kopf run -m <whatever handler module>
```
</details>

<details><summary>The full output of the command that failed</summary>

```
Usage: kopf run [OPTIONS] [PATHS]...                                                                                                                                                                                                                          
Try 'kopf run --help' for help.                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                              
Error: Invalid value for '--log-format': <LogFormat.FULL: '[%(asctime)s] %(name)-20.20s [%(levelname)-8.8s] %(message)s'> is not one of 'plain', 'full', 'json'. 
```
</details>


## Environment

<!-- The following commands can help:
     `kopf --version` or `pip show kopf`
     `kubectl version`
     `python --version`
-->

* Kopf version: 1.31.0
* Kubernetes version:  1.18.14
* Python version: 3.8
* OS/platform: Ubuntu 20.04

<details><summary>Python packages installed</summary>
<!-- use `pip freeze --all` -->
relevant package installed is

```
click==8.0.0
```
</details>

Immediate workaround is to add explicitly `--log-format full` to the `kopf run` invocation

**Comment by pshchelo** [comment]
let's hold it for now. I get really strange results when installing in a clean docker container, in a clean virtualenv and as pip install -e . in another virtualenv
sometimes it works, sometimes it does not. sometimes old code works with click 8.x, sometimes it does not

need to investigate further. I'd really like someone else to chime in if they see the same problem

**Comment by nolar** [comment]
@pshchelo I agree. Should we now restrict to click<8.0.0 until it is fixed? So that the initial journey could work; so as Kopf upgrades to new releases in other people’s apps.

**Comment by pshchelo** [comment]
@nolar seems like a least intrusive immediate option at the moment

**Comment by SStorm** [comment]
Can confirm this is happening and that click 8.0.0 seems incompatible. Please can we fix versions of all dependencies to avoid this happening in the future?

**Comment by SStorm** [comment]
On a related note, this not only breaks the latest kopf release, but all previous kopf releases too :(

We worked around it in our app by fixing the version of `click<8.0.0` next to where the kopf dependency is specified.

**Comment by nolar** [comment]
Explained and fixed in #769 (investigated in #768).

Released in [1.32rc1](https://github.com/nolar/kopf/releases/tag/1.32rc1) (beware: there are other changes and massive code rebalancing — see the changelog).

**As an immediate workaround for bleeding cases when Kopf upgrade is not possible and older versions are required, restrict `click<8.0.0` in your requirements.**

There will be no backports to 1.31.x or older versions — use the work-around instead.

**Comment by nolar** [comment]
@SStorm Kopf is a library. It does not fix or pin the _runtime dependencies_ (`setup.py`) and barely restricts them except as via `>=` (where new features of existing dependencies are needed — only to trigger the cascaded upgrades). As a library, it should give freedom to the application developers to decide which versions of all dependencies they use (combined with their sophisticated security and compliance policies, environments, integrations with other libraries, etc.). 

It is generally advised for the applications (operators) to pin their versions of **ALL** dependencies (incl. Kopf), but this is out of scope of Kopf and is a responsibility of the application authors/maintainers. In our work team, we do this with [`pip-tools/pip-compile`](https://github.com/jazzband/pip-tools); Poetry's and Pipenv's lockfiles are also an option. This would prevent the issue with Kopf + Click 8.0.0.

As for Kopf's _CI/test-time dependencies_ (`requirements.txt`), pinning can be considered there. Though, I prefer to "check" Kopf's compatibility (and to "detect" Kopf's incompatibility) with the newer versions of everything by failing the CI — rather than by introducing excessive automation by e.g. dependabot. That is a debatable approach, but it works "good enough" (this might change when the circumstances change).
